wagtail/contrib/search_promotions/views.py

```diff
@@ -6,80 +6,74 @@
+from wagtail.admin.views import generic
 from wagtail.contrib.search_promotions import forms, models
-from wagtail.contrib.search_promotions.models import Query
+from wagtail.contrib.search_promotions.models import Query, SearchPromotion
 from wagtail.log_actions import log
+from wagtail.permission_policies.base import ModelPermissionPolicy
 from wagtail.search.utils import normalise_query_string
 
 
-@any_permission_required(
-    "wagtailsearchpromotions.add_searchpromotion",
-    "wagtailsearchpromotions.change_searchpromotion",
-    "wagtailsearchpromotions.delete_searchpromotion",
-)
-@vary_on_headers("X-Requested-With")
-def index(request):
-    # Ordering
-    valid_ordering = ["query_string", "-query_string", "views", "-views"]
-    ordering = valid_ordering[0]
-
-    if "ordering" in request.GET and request.GET["ordering"] in valid_ordering:
-        ordering = request.GET["ordering"]
-
-    # Query
-    queries = Query.objects.filter(editors_picks__isnull=False).distinct()
-
-    if "views" in ordering:
-        queries = queries.annotate(views=functions.Coalesce(Sum("daily_hits__hits"), 0))
-
-    queries = queries.order_by(ordering)
-
-    # Search
-    is_searching = False
-    query_string = request.GET.get("q", "")
-
-    if query_string:
-        queries = queries.filter(query_string__icontains=query_string)
-        is_searching = True
-
-    # Paginate
-    paginator = Paginator(queries, per_page=20)
-    try:
-        queries = paginator.page(request.GET.get("p", 1))
-    except InvalidPage:
-        raise Http404
-
-    if request.headers.get("x-requested-with") == "XMLHttpRequest":
-        return TemplateResponse(
-            request,
-            "wagtailsearchpromotions/results.html",
-            {
-                "is_searching": is_searching,
-                "ordering": ordering,
-                "queries": queries,
-                "query_string": query_string,
-            },
-        )
-    else:
-        return TemplateResponse(
-            request,
-            "wagtailsearchpromotions/index.html",
-            {
-                "is_searching": is_searching,
-                "ordering": ordering,
-                "queries": queries,
-                "query_string": query_string,
-                "search_form": SearchForm(
-                    data={"q": query_string} if query_string else None,
-                ),
-            },
-        )
+class SearchPromotionColumn(RelatedObjectsColumn):
+    cell_template_name = "wagtailsearchpromotions/search_promotion_column.html"
+
+
+class IndexView(generic.IndexView):
+    model = Query
+    template_name = "wagtailsearchpromotions/index.html"
+    results_template_name = "wagtailsearchpromotions/index_results.html"
+    context_object_name = "queries"
+    page_title = gettext_lazy("Search Terms")
+    header_icon = "pick"
+    paginate_by = 20
+    permission_policy = ModelPermissionPolicy(SearchPromotion)
```

Comment:
I was a bit surprised that there were no failing test cases when we didn't have this, so we likely need some tests to ensure that users without `add`/`change`/`delete` permissions for the `SearchPromotion` model should not be able to access the view.

Let me know if you'd like to try adding the test, otherwise I can do it as I merge 😃

Reply:
@laymonage I've added the test for permissions. please let me know if something is needed in it. I'm also trying to add tests for the views column.